src/lib.rs

```diff
@@ -1674,6 +1674,77 @@ impl<B: BitBlock> BitVec<B> {
             carry = curr_carry;
         }
     }
+
+    // todo(jujumba): docs
+    pub fn insert_many(&mut self, at: usize, bits: &[bool]) {
+        assert!(
+            at <= self.nbits,
+            "insertion index (is {at}) should be <= nbits (is {nbits})",
+            nbits = self.nbits
+        );
+
+        let block_at = at / B::bits();
+        let bit_at = at % B::bits();
+        let block_offset = bits.len() / B::bits();
+
+        self.extend_if_needed(self.nbits + bits.len());
+
+        unsafe {
+            self.rotate(at, bits.len() % B::bits());
+        }
+
+        for i in ((block_at + 1)..=(block_at + block_offset)).rev() {
+            let block = core::mem::replace(&mut self.storage[i], B::zero());
+            self.storage[i + block_offset] = block;
```

Comment:
Do you mean something like this?

```rust
for i in ((block_at + 1)..=(block_at + block_offset)).rev() {
    let block = core::mem::replace(&mut self.storage[i], B::zero());
    let carry = block >> (B::bits() - offset);
    self.storage[i + block_offset] = block >> offset;
    self.storage[i + block_offset + 1] = self.storage[i + block_offset + 1] | carry;
}
````

Reply:
Almost. You can get carry from the previous iteration, not the current iteration. Then the change to `self.storage[i + block_offset + 1]` will wait until another iteration.

Reply:
The change to `self.storage[i + block_offset + 1]` is not really needed

Reply:
> Almost. You can get carry from the previous iteration, not the current iteration. Then the change to `self.storage[i + block_offset + 1]` will wait until another iteration.

And then I have to start from `block_at +1`, right?

I'm intentionally starting from the next block, because the last bits in the first are tricky, and may get to different blocks, shifting them is not so trivial (though there is a decent chance I'm wrong)

Reply:
@pczarn , any thoughts?..

Reply:
@jujumba Sorry for the delay.

Yes, you'd start at another block and perhaps finish sooner. Keep in mind we are going in reverse, rather than

But, I would just load the carry from past the destination, then go through ALL the blocks without skipping any, then handle the carry to store the beginning of the source.

Also, looks like you have block_offset mixed up, adding it twice. Did the tests pass the last time?

Reply:
> But, I would just load the carry from past the destination, then go through ALL the blocks without skipping any, then handle the carry to store the beginning of the source.

I am not 100% sure that I get what you mean, but I will try to get down to it in a day or two

Reply:
> Do you mean something like this?
> 
> ```rust
> let offset = bits.len() % B::bits();
> for i in ((block_at + 1)..=(block_at + block_offset)).rev() {
>     let block = core::mem::replace(&mut self.storage[i], B::zero());
>     let carry = block >> (B::bits() - offset);
>     self.storage[i + block_offset] = block >> offset;
>     self.storage[i + block_offset + 1] = self.storage[i + block_offset + 1] | carry;
> }
> ```

Actually that seems incorrect, the loop has to start from `i + block_offset + 1` and run till the last block (before extension the storage with `extend_if_needed`

Reply:
> Almost. You can get carry from the previous iteration, not the current iteration. Then the change to `self.storage[i + block_offset + 1]` will wait until another iteration.

Do you mean something like this?
```rust
let mut carry = B::zero();
for i in ((block_at + block_offset + 1)..=len).rev() {
    let block = core::mem::replace(&mut self.storage[i], B::zero());
    let curr_carry = block >> (B::bits() - offset);
    self.storage[i + 1] = self.storage[i + 1] | carry;
    self.storage[i + block_offset] = block >> offset;
    carry = curr_carry;
}
```
I am not quite sure that this is possible in the first place, as to rotate bits in blocks by `offset` we have to go in "direct" order (that is, not in reverse), store the carry from the current iteration and reuse it in the one. However, to shift blocks (just reassign them by `block_offset`), we have to go in reverse order.

Also this code should fail if we are rotating and shifting by one. Starting from the last block, we capture the first bit. Then assign that bit to the next block. And then assign `block >> offset` to the next block (that is, we have overwritten the carry).

So, could you, please, provide some clues in code?

@pczarn

Reply:
I can't dedicate that much time this week :( maybe Saturday or Monday

Reply:
@jujumba I can't really do that much for the next month. Though the progress we did is pretty solid.

What's more, my laptop for development has died.

Reply:
> @Jujumba I can't really do that much for the next month. Though the progress we did is pretty solid.
> 
> What's more, my laptop for development has died.

Okay, then I'm waiting for you 🫡

Reply:
> > Almost. You can get carry from the previous iteration, not the current iteration. Then the change to `self.storage[i + block_offset + 1]` will wait until another iteration.
> 
> Do you mean something like this?
> 
> ```rust
> let mut carry = B::zero();
> for i in ((block_at + block_offset + 1)..=len).rev() {
>     let block = core::mem::replace(&mut self.storage[i], B::zero());
>     let curr_carry = block >> (B::bits() - offset);
>     self.storage[i + 1] = self.storage[i + 1] | carry;
>     self.storage[i + block_offset] = block >> offset;
>     carry = curr_carry;
> }
> ```
> 
> I am not quite sure that this is possible in the first place, as to rotate bits in blocks by `offset` we have to go in "direct" order (that is, not in reverse), store the carry from the current iteration and reuse it in the one. However, to shift blocks (just reassign them by `block_offset`), we have to go in reverse order.
> 
> Also this code should fail if we are rotating and shifting by one. Starting from the last block, we capture the first bit. Then assign that bit to the next block. And then assign `block >> offset` to the next block (that is, we have overwritten the carry).
> 
> So, could you, please, provide some clues in code?
> 
> @pczarn

Well, I just looked at this code again, and I'm pretty sure that it is incorrect. I guess the simplest solution (yet still effective), is to just rotate by `bits.len() % B::bits()` starting at `at`, and then shift all blocks starting with block `bits.len() / B::bits() + 1` by `bits.len() / B::bits()`

Your thoughts?

@pczarn (sorry for pinging you :innocent: )

Reply:
> > > Almost. You can get carry from the previous iteration, not the current iteration. Then the change to `self.storage[i + block_offset + 1]` will wait until another iteration.
> > 
> > 
> > Do you mean something like this?
> > ```rust
> > let mut carry = B::zero();
> > for i in ((block_at + block_offset + 1)..=len).rev() {
> >     let block = core::mem::replace(&mut self.storage[i], B::zero());
> >     let curr_carry = block >> (B::bits() - offset);
> >     self.storage[i + 1] = self.storage[i + 1] | carry;
> >     self.storage[i + block_offset] = block >> offset;
> >     carry = curr_carry;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am not quite sure that this is possible in the first place, as to rotate bits in blocks by `offset` we have to go in "direct" order (that is, not in reverse), store the carry from the current iteration and reuse it in the one. However, to shift blocks (just reassign them by `block_offset`), we have to go in reverse order.
> > Also this code should fail if we are rotating and shifting by one. Starting from the last block, we capture the first bit. Then assign that bit to the next block. And then assign `block >> offset` to the next block (that is, we have overwritten the carry).
> > So, could you, please, provide some clues in code?
> > @pczarn
> 
> Well, I just looked at this code again, and I'm pretty sure that it is incorrect. I guess the simplest solution (yet still effective), is to just rotate by `bits.len() % B::bits()` starting at `at`, and then shift all blocks starting with block `bits.len() / B::bits() + 1` by `bits.len() / B::bits()`
> 
> Your thoughts?
> 
> @pczarn (sorry for pinging you 😇 )

Or perhaps this code is better:
```rust
        for i in ((block_at + 1)..=len).rev() {
            let block = core::mem::replace(&mut self.storage[i], B::zero());
            let curr_carry = block >> (B::bits() - offset);
            self.storage[i + block_offset] = self.storage[i + block_offset] | carry;
            self.storage[i + block_offset] = block << offset;
            carry = curr_carry;
        }
```
But it doesn't handle bits in the `block_at`, as we can't just shift it around, so the second loop is necessary in any way

Reply:
> > > > Almost. You can get carry from the previous iteration, not the current iteration. Then the change to `self.storage[i + block_offset + 1]` will wait until another iteration.
> > > 
> > > 
> > > Do you mean something like this?
> > > ```rust
> > > let mut carry = B::zero();
> > > for i in ((block_at + block_offset + 1)..=len).rev() {
> > >     let block = core::mem::replace(&mut self.storage[i], B::zero());
> > >     let curr_carry = block >> (B::bits() - offset);
> > >     self.storage[i + 1] = self.storage[i + 1] | carry;
> > >     self.storage[i + block_offset] = block >> offset;
> > >     carry = curr_carry;
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I am not quite sure that this is possible in the first place, as to rotate bits in blocks by `offset` we have to go in "direct" order (that is, not in reverse), store the carry from the current iteration and reuse it in the one. However, to shift blocks (just reassign them by `block_offset`), we have to go in reverse order.
> > > Also this code should fail if we are rotating and shifting by one. Starting from the last block, we capture the first bit. Then assign that bit to the next block. And then assign `block >> offset` to the next block (that is, we have overwritten the carry).
> > > So, could you, please, provide some clues in code?
> > > @pczarn
> > 
> > 
> > Well, I just looked at this code again, and I'm pretty sure that it is incorrect. I guess the simplest solution (yet still effective), is to just rotate by `bits.len() % B::bits()` starting at `at`, and then shift all blocks starting with block `bits.len() / B::bits() + 1` by `bits.len() / B::bits()`
> > Your thoughts?
> > @pczarn (sorry for pinging you 😇 )
> 
> Or perhaps this code is better:
> 
> ```rust
>         for i in ((block_at + 1)..=len).rev() {
>             let block = core::mem::replace(&mut self.storage[i], B::zero());
>             let curr_carry = block >> (B::bits() - offset);
>             self.storage[i + block_offset] = self.storage[i + block_offset] | carry;
>             self.storage[i + block_offset] = block << offset;
>             carry = curr_carry;
>         }
> ```
> 
> But it doesn't handle bits in the `block_at`, as we can't just shift it around, so the second loop is necessary in any way

After a bit of thinking I came to this:
```rust
        for i in (block_at..=len).rev() {
            // last iteration, handle the tricky bits :)
            if i == block_at {
                let curr_carry = self.storage[i] >> (B::bits() - offset);
                self.storage[i + block_offset] = curr_carry;
                continue;
            }
            let block = core::mem::replace(&mut self.storage[i], B::zero());
            let curr_carry = block >> (B::bits() - offset);
            self.storage[i + block_offset] = self.storage[i + block_offset] | carry;
            self.storage[i + block_offset] = block << offset;
            carry = curr_carry;
        }
```

You just have to move bits after `bit_at` at `block_at` to the `block_at + block_offset` block.

Reply:
> Hey,
> 
> I am back online.
> 
> > (sorry for pinging you 😇 )
> 
> My phone died as well, so no pings could get to me regardless. :(

And are you back to maintaining the crate? If so, could you have a look on my thoughts?

Reply:
@Jujumba You cannot go to `..= len`, you must remain in bounds. Though I bet it is possible to make a good trick there by extending by one block.

I started looking into this.

Reply:
The code you posted most recently must have a mistake, because you are overwriting what you did with `self.storage[foo] | carry`.

Reply:
Going in reverse is a good idea. And, modern processors handle reverse iteration as good as normal forward iteration.

I will try to dig deeper into your code within a few days.

Reply:
Hey, I'm a bit busy at my job and have some other stuff going on. I didn't lost interest in closing this and will eventually return

Reply:
Well, got down to it again, and there  are so many corner cases which makes the implementation of this function extremely hard :(

Reply:
@Jujumba Okay, I've been thinking whether better give advice or finish myself. Seems like some complex stuff, I should get it working within a few days.

Reply:
@Jujumba With algorithms like these, often there are tricks we may come up with to deal with edge cases effectively and make our work simpler.